tutorials/arvo/cli-tutorial.md

```diff
@@ -0,0 +1,427 @@
+generators, which only covers the bare minimum necessary to write generators
+that take user input.
+
+
+## The `shoe` library {#the-shoe-library}
+
+Here we describe how sessions are identified, the `%shoe` `card`s that Gall agents
+with the `shoe` library are able to utilize, and the different cores of `/lib/shoe.hoon` and their purpose.
+
+### Session identifiers
+
+An app using the `shoe` library will track sessions with an opaque
+`sole-ids=(list @ta)`. These are typically called session ids, are are
+opaque identifiers generated by the connecting client. An app using the `shoe`
+library may be connected to by a local or remote ship in order to send commands,
+and each of these connections is assigned a unique `@ta` that identifies the
+ship and which session on that ship if there are multiple.
+
+### `%shoe` cards
+
+Gall agents with the `shoe` library are able to utilize `%shoe` `card`s. These
+have the following shape.
+```hoon
+[%shoe sole-ids=(list @ta) effect=shoe-effect]`
+```
+The `sole-ids` is the `list` of session ids that the following `effect` is
+emitted to. An empty `sole-ids` sends the effect to all connected sessions.
+`shoe-effect`s, for now, are always of the shape `[%sole effect=sole-effect]`, where
+`sole-effect`s are basic console events such as displaying text, changing the
+prompt, beeping, etc. These are described in the section on the [`sole` library](#the-sole-library).
+
+For example, a `%shoe` `card` that causes all connected sessions to beep would
+be `[%shoe ~ %sole %bel ~]`. 
+
+### `shoe` core
+
+An iron (contravariant) door that defines an interface for Gall agents utilizing
+the `shoe` library. Use this core whenever you want to receive input from the user and run a command. The input will get
+put through the parser (`+command-parser`) and results in a noun of
+`command-type` that the underlying application specifies, which shoe then feeds back into the underlying app as an `+on-command` callback.
+
+In addition to the ten arms that all Gall core apps possess, `+shoe` has a few
+more specific to making use of the `shoe` library. Thus you will need to wrap the `shoe:shoe` core with the `agent:shoe` core to obain a standard
+10-arm Gall agent core. See the [shoe example app
+walkthrough](#shoe-example-app-walkthrough) for how to do this.
+
+Under each header we give the type signature for the arm.
+
+#### `+command-parser`
+
+```hoon
+  ++  command-parser
+    |~  sole-id=@ta
+    |~(nail *(like [? command-type]))
+```
+
+Input parser for a specific command-line session. If the head of the result is true,
+instantly run the command. If it's false, require the user to press return.
+
+#### `+tab-list`
+
+```hoon
+  ++  tab-list
+    |~  sole-id=@ta
+    *(list (option:auto tank))
+```
+
+Autocomplete options for the command-line session (to match `+command-parser`).
+
+#### `+on-command`
+
+```hoon
+  ++  on-command
+    |~  [sole-id=@ta command=command-type]
+    *(quip card _^|(..on-init))
```

Comment:
Wondering if we should lie and say `_this` in place of `_^|(..on-init)` instead. It's semantically equivalent, and more likely to be familiar. The current thing suffers from lack of context.

Reply:
Yeah I have to think for a moment to see why they're the same. `_this` is much better.